source/agora/network/Crawler.d

```diff
@@ -0,0 +1,526 @@
+
+    /// Clock
+    private Clock clock;
+
+    /// 1. Crawler fibers pop an already dicovered node from the front of the list.
+    /// 2. Crawler fibers try to get node recommendations from the popped node.
+    /// 3. If the popped node answers with a list of node recommendations,
+    ///    then all those new recommendations and the popped node are pushed to
+    ///    the back of the list.
+    private DList!NetworkClientHolder discovered_node_list;
+
+    /// Lookup map to decide, whether the current node has already discovered
+    /// a particular node. This maps supplements `discovered_node_list` and
+    /// was added because:
+    ///
+    /// 1. Iterating over `discovered_node_list` is slow.
+    /// 2. Iterating over `discovered_node_list` cannot reliably answer, whether
+    ///    a node has been discovered, as crawling fibers pop elements off of it.
+    private NetworkClientHolder[Address] discovered_node_map;
+
+    /// Contains the crawling results. A particular node is removed from this
+    /// data structure, after it is temporarily banned.
+    private CrawlResultHolder crawl_result_holder;
+
+    /// Indicating whether the crawler is in the process of shutting down.
+    private bool is_shutting_down;
+
+    /// Number of crawler fibers that already shut down.
+    private ubyte already_shutdown_cnt;
+
+    public this (ITaskManager taskman, Clock clock, Config config, NetworkManager network_manager)
+    {
+        this.taskman = taskman;
+        this.clock = clock;
+        this.config = config;
+        this.network_manager = network_manager;
+        this.node_locator = this.makeNodeLocator();
+        this.log = Logger(__MODULE__);
+
+        BanManager.Config ban_config =
+        {
+            max_failed_requests : 25,
+            ban_duration : 30.minutes
+        };
+        this.banman = new BanManager(ban_config, clock, "banned_crawler.dat");
+    }
+
+    /// Entry point for the crawling fibers
+    private void crawl ()
+    {
+        // Fiber shutdown have to be coordinated among all fibers
+        while (!this.is_shutting_down)
+        {
+            if (discovered_node_list.empty)
+            {
+                this.taskman.wait(this.config.node.crawling_interval);
+                continue;
+            }
+
+            // Pop a node from the front of the list
+            auto network_client_holder = this.discovered_node_list.front;
+            auto client_address = network_client_holder.network_client.address;
+            this.discovered_node_list.removeFront();
+
+            NodeInfo node_info;
+            try
+                // Node might be banned because of too many wrong recommendation,
+                // despite the fact the node is constantly available
+                if (this.banman.isBanned(client_address))
+                    throw new Exception("Client banned");
+                else
+                    node_info = network_client_holder.network_client.getNodeInfo();
```

Comment:
A nested function sounds more reasonable than throwing an `Exception` to follow the same code path.